src/controllers/hid/hidiothread.cpp

```diff
@@ -0,0 +1,300 @@
+                        << m_deviceInfo.formatName() << "serial #"
+                        << m_deviceInfo.serialNumber()
+                        << "(including one byte for the report ID:"
+                        << QString::number(static_cast<quint8>(reportID), 16)
+                                   .toUpper()
+                                   .rightJustified(2, QChar('0'))
+                        << ") - Needed: "
+                        << (mixxx::Time::elapsed() - startOfHidGetInputReport)
+                                   .formatMicrosWithUnit();
+
+    if (bytesRead <= kReportIdSize) {
+        // -1 is the only error value according to hidapi documentation.
+        // Otherwise minimum possible value is 1, because 1 byte is for the reportID,
+        // the smallest report with data is therefore 2 bytes.
+        DEBUG_ASSERT(bytesRead <= kReportIdSize);
+        return QByteArray();
+    }
+
+    // Convert array of bytes read in a JavaScript compatible return type, this is returned as deep-copy, for thread safety.
+    // For compatibility with HidController::processInputReport, the reportID prefix is included added here
+    return QByteArray(
+            reinterpret_cast<char*>(m_pPollData[m_pollingBufferIndex]),
+            bytesRead);
+}
+
+void HidIoThread::latchOutputReport(const QByteArray& data, unsigned int reportID) {
+    {
+        auto lock = lockMutex(&m_outputReportMapMutex);
+        if (m_outputReports.find(reportID) == m_outputReports.end()) {
+            std::unique_ptr<HidIoReport> pNewOutputReport;
+            m_outputReports[reportID] = std::make_unique<HidIoReport>(
+                    reportID);
+        }
+    }
+    m_outputReports[reportID]->latchOutputReport(data, m_deviceInfo, m_logOutput);
+}
+
+bool HidIoThread::sendNextOutputReport() {
+    // m_outputReports.size() doesn't need mutex protection, because the value of i is not used.
+    // i is just a counter to prevent infinite loop execution.
+    // If the map size increases, this loop will execute one iteration more,
+    // which only has the effect, that one additional lookup operation for unsend data will be executed.
+    for (unsigned char i = 0; i < m_outputReports.size(); i++) {
+        {
+            auto lock = lockMutex(&m_outputReportMapMutex);
+            m_outputReportIterator++;
+            if (m_outputReportIterator == m_outputReports.end()) {
+                m_outputReportIterator = m_outputReports.begin();
+            }
+        }
+
+        auto lock = lockMutex(&m_hidDeviceMutex);
+
+        // The only operator used in this class to modify the map container itself,
+        // is insert by std::map<Key,T,Compare,Allocator>::operator[]
+        // The standard says that "No iterators or references are invalidated." using this operator.
+        // Therefore m_outputReportIterator doesn't require Mutex protection.
+        if (m_outputReportIterator->second->sendOutputReport(
```

Comment:
m_outputReportIterator might have been modified when getting here after unlocking m_outputReportMapMutex above. It could even be invalidated between reading and dereferencing it. I don't see how this could be thread-safe.

Reply:
If entries are only ever added and never removed than this might actually work. But then it has to be properly documented. You mentioned `insert` but not the fact the this is the only mutable operation on `m_outputReports`.

Reply:
I changed the wording of the comment and added another to describe the function of the map.